druid/src/menu/mod.rs

```diff
@@ -0,0 +1,762 @@
+    // above, but "static" versions of them not depending on the data.
+    old_state: Option<MenuItemState>,
+}
+
+impl<T: Data> From<MenuItem<T>> for MenuEntry<T> {
+    fn from(i: MenuItem<T>) -> MenuEntry<T> {
+        MenuEntry { inner: Box::new(i) }
+    }
+}
+
+struct Separator;
+
+impl<T: Data> From<Separator> for MenuEntry<T> {
+    fn from(s: Separator) -> MenuEntry<T> {
+        MenuEntry { inner: Box::new(s) }
+    }
+}
+
+impl<T: Data> Menu<T> {
+    /// Create an empty menu.
+    pub fn empty() -> Menu<T> {
+        Menu {
+            needs_rebuild: None,
+            needs_refresh: None,
+            item: MenuItem::new(""),
+            children: Vec::new(),
+        }
+    }
+
+    /// Create a menu with the given name.
+    pub fn new(title: impl Into<LabelText<T>>) -> Menu<T> {
+        Menu {
+            needs_rebuild: None,
+            needs_refresh: None,
+            item: MenuItem::new(title),
+            children: Vec::new(),
+        }
+    }
+
+    /// Provide a callback for determining whether this item should be enabled.
+    ///
+    /// Whenever the callback returns `true`, the menu will be enabled.
+    pub fn enabled(mut self, enabled: impl FnMut(&T, &Env) -> bool + 'static) -> Self {
+        self.item = self.item.enabled(enabled);
+        self
+    }
+
+    /// Append a menu entry to this menu, returning the modified menu.
+    pub fn with_entry(mut self, entry: impl Into<MenuEntry<T>>) -> Self {
+        self.append_entry(entry);
+        self
+    }
+
+    /// Append a menu entry to this menu.
+    pub fn append_entry(&mut self, entry: impl Into<MenuEntry<T>>) {
+        self.children.push(entry.into());
+    }
+
+    /// Append a separator to this menu, returning the modified menu.
+    pub fn with_separator(mut self) -> Self {
+        self.append_separator();
+        self
+    }
+
+    /// Append a separator to this menu.
+    pub fn append_separator(&mut self) {
+        self.children.push(Separator.into());
+    }
+
+    /// Supply a function to check when this menu needs to refresh itself.
+    ///
+    /// The arguments to the callback are (in order):
+    /// - the previous value of the data,
+    /// - the current value of the data, and
+    /// - the current value of the environment.
+    ///
+    /// The callback should return true if the menu needs to refresh itself.
+    ///
+    /// This callback is intended to be purely an optimization. If you do create a menu without
+    /// supplying a refresh callback, the menu will recursively check whether any children have
+    /// changed and refresh itself if any have. By supplying a callback here, you can short-circuit
+    /// those recursive calls.
+    pub fn needs_refresh(mut self, refresh: impl FnMut(&T, &T, &Env) -> bool + 'static) -> Self {
+        self.needs_refresh = Some(Box::new(refresh));
+        self
+    }
+
+    /// Supply a function to check when this menu needs to be rebuild from scratch.
```

Comment:
```suggestion
    /// Supply a function to check when this menu needs to be rebuilt from scratch.
```

Reply:
I'm just gonna do this in my local copy to avoid an extra rebase